# Heart Attack Prediction (60pt)

In [1]:
# Load the necessary packages
import numpy as np
import pandas as pd
import statsmodels.formula.api as smf
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import mean_squared_error
from sklearn.metrics import accuracy_score
from sklearn.metrics import precision_score
from sklearn.metrics import recall_score
from sklearn.metrics import confusion_matrix

    In this question, we will construct a simple logistic regression model to predict the probability of a person having a heart attack. The dataset comes from Kaggle www.kaggle.com/rashikrahmanpritom/heart-attack-analysis-prediction-dataset, which contains health information of each person (predictors) and whether or not the person had a heart attack before (outcome variable). You can download the data heart.csv from Canvas.

    1. (2pt) Load the data and drop the variable slp, oldpeak and thall, since they do not have descriptions available. Do some basic sanity check (descriptive statistics; missing values; data types...). The data should contain 303 rows, and 11 columns.

In [2]:
# Read the heart attack dataframe
heart_df = pd.read_csv('/home/jovyan/PS/data/heart.csv', sep=',')
# Drop slp, oldpeak and thall columns
heart_df.drop(['slp', 'oldpeak', 'thall'], axis = 1, inplace = True)
# Sanity Checks
heart_df.info()
heart_df.head(5)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 11 columns):
 #   Column    Non-Null Count  Dtype
---  ------    --------------  -----
 0   age       303 non-null    int64
 1   sex       303 non-null    int64
 2   cp        303 non-null    int64
 3   trtbps    303 non-null    int64
 4   chol      303 non-null    int64
 5   fbs       303 non-null    int64
 6   restecg   303 non-null    int64
 7   thalachh  303 non-null    int64
 8   exng      303 non-null    int64
 9   caa       303 non-null    int64
 10  output    303 non-null    int64
dtypes: int64(11)
memory usage: 26.2 KB


age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  caa  output
0   63    1   3     145   233    1        0       150     0    0       1
1   37    1   2     130   250    0        1       187     0    0       1
2   41    0   1     130   204    0        0       172     0    0       1
3   56    1   1     120   236    0        1       178     0    0       1
4   57    0   0     120   354    0        1       163     1    0       1

    2. (22pt) Construct a simple logistic regression with statsmodel.formula.api package. The outcome variable is output. The rest 10 variables (age, sex, cp, trtbps, chol, fbs, restecg, thalachh, exng, caa) in the dataset are predictors. Print out the marginal effect summary table and answer the following questions:

In [3]:
# Construct the logistic regression and print the marginal effect summary table
logit_regression = smf.logit("output ~ age + sex + C(cp) + trtbps + chol + fbs + C(restecg) + thalachh + exng + caa", data = heart_df).fit()
logit_regression.get_margeff().summary()

Optimization terminated successfully.
         Current function value: 0.386917
         Iterations 7


<class 'statsmodels.iolib.summary.Summary'>
"""
        Logit Marginal Effects       
=====================================
Dep. Variable:                 output
Method:                          dydx
At:                           overall
===================================================================================
                     dy/dx    std err          z      P>|z|      [0.025      0.975]
-----------------------------------------------------------------------------------
C(cp)[T.1]          0.2017      0.060      3.353      0.001       0.084       0.320
C(cp)[T.2]          0.2278      0.045      5.084      0.000       0.140       0.316
C(cp)[T.3]          0.2117      0.071      2.977      0.003       0.072       0.351
C(restecg)[T.1]     0.0604      0.042      1.452      0.146      -0.021       0.142
C(restecg)[T.2]    -0.1007      0.216     -0.467      0.641      -0.523       0.322
age                -0.0007      0.003     -0.269      0.788      -0.006       0.005
sex                -0.2439      0.047     -5.233      0.000      -0.335      -0.153
trtbps             -0.0025      0.001     -2.116      0.034      -0.005      -0.000
chol               -0.0008      0.000     -1.866      0.062      -0.002    4.18e-05
fbs                 0.0267      0.059      0.449      0.654      -0.090       0.143
thalachh            0.0039      0.001      3.426      0.001       0.002       0.006
exng               -0.1354      0.046     -2.943      0.003      -0.226      -0.045
caa                -0.0961      0.018     -5.213      0.000      -0.132      -0.060
===================================================================================
"""

In this model, only cp and restecg are seen as categorical variables. Both "cp" and "restecg" have more than two categories, and these are clearly not interval measures.  So these are inherently categorical, and you should convert those to categories. Sex is categorical. However, it is normally coded as 0/1 (as it did in this dataset), so it does not need to be seen as categorical or make a dummy out of it.

    (6pt) Interpret coefficient for sex. Is it statistically significant at significance level of 0.05?

Males have 24.39 percentage points less chance to get a heart attack compared to females if the other characters are the same. The coefficient is statistically significant because the p-value is less than the significance level of 0.05.

    (6pt) Interpret coefficient for cp. Is it statistically significant at significance level of 0.05?    

When other characteristics are the same, a patient that has atypical angina chest have 20.17 percentage points more likelihood to get a heart attack, a patient that has non-anginal pain have 22.78 percentage points more likelihood to get a heart attack, a patient that has asymptomatic chest pain have 21.17 percentage points more likelihood to get a heart attack. All these are compared with a patient that has typical angina chest pain. 

The coefficient is statistically significant as the p-value is less than the significance level of 0.05.

    (6pt) Interpret the coefficient for age. Is it statistically significant at significance level of 0.05?

When other characteristics are the same, with one unit increase in age, the chance of getting heart attacks decreases by 0.07 percentage points. 

The coefficient is not statistically significant as the p-value is way larger than the significance level of 0.05.

    (4pt) What are the variables that are associated with lower chance of having a heart attack, and also are statistically significant? Do they intuitively make sense? Why?

Variables that are associated with a lower chance of having a heart attack and also statistically significant are resting blood pressure(trtbps), exercise-induced angina(exng), the number of major vessels(caa), and male patients(sex = 1). However, they do not intuitively makes sense because all these variables are associated with a higher chance of having a heart attack based on research and common sense (i.e. having angina increases the chance of getting a heart attack).

    (6pt) Now let’s construct the same model with sklearn package using the following code:
    LogisticRegression(penalty=’none’, solver=’newton-cg’).fit(X,y)
    Remember that sklearn package requires the predictor values (X) to be separated with the outcome variable (y). The predictor values also need to be in matrix format. Answer the following question:

In [4]:
# Crtaed the predictor values and the outcome variables
X1 = heart_df[["age", "sex", "cp", "trtbps", "chol", "fbs", "restecg", "thalachh", "exng", "caa"]]
X = pd.get_dummies(X1, drop_first = True, columns = ["cp", "restecg"]).values
y = heart_df["output"].values
# Create a logistic regression model with sklearn
sklearn_logistic = LogisticRegression(penalty = 'none', solver = 'newton-cg').fit(X,y)

    3.(6pt) Print out the coefficient and intercept. Did you get the same intercept and coefficients as what you got from statsmodel.formula.api package? (you are supposed to!!)
    (After using both packages, you should have a clearer sense that smf package is good for inferences becuase it provides information about pvalues, tvalues, CI etc., which can be used to determine statistical significance. Sklean package does not provide such information. However, the advantage of sklearn is that it is easier to construct models and make prediction with sklearn, especially when you are comparing multiple models.)

In [5]:
# Print the coefficients and intercept of the sklearn model
print('Coefficients:', sklearn_logistic.coef_)
print('Intercept:', sklearn_logistic.intercept_)

Coefficients: [[-0.005895   -1.98329754 -0.02046839 -0.0067561   0.21704038  0.03143576
  -1.10053597 -0.78101777  1.63967367  1.8519962   1.72131152  0.49135353
  -0.81831285]]
Intercept: [1.27592977]


In [6]:
# Print the summary of the smf logistic regression summary
logit_regression.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                 output   No. Observations:                  303
Model:                          Logit   Df Residuals:                      289
Method:                           MLE   Df Model:                           13
Date:                Thu, 17 Mar 2022   Pseudo R-squ.:                  0.4386
Time:                        18:52:45   Log-Likelihood:                -117.24
converged:                       True   LL-Null:                       -208.82
Covariance Type:            nonrobust   LLR p-value:                 3.830e-32
===================================================================================
                      coef    std err          z      P>|z|      [0.025      0.975]
-----------------------------------------------------------------------------------
Intercept           1.2759      2.326      0.548      0.583      -3.283       5.835
C(cp)[T.1]          1.6397      0.521      3.147      0.002       0.618       2.661
C(cp)[T.2]          1.8520      0.417      4.446      0.000       1.036       2.668
C(cp)[T.3]          1.7213      0.608      2.833      0.005       0.530       2.912
C(restecg)[T.1]     0.4914      0.342      1.437      0.151      -0.179       1.162
C(restecg)[T.2]    -0.8183      1.755     -0.466      0.641      -4.258       2.622
age                -0.0059      0.022     -0.269      0.788      -0.049       0.037
sex                -1.9833      0.430     -4.616      0.000      -2.825      -1.141
trtbps             -0.0205      0.010     -2.060      0.039      -0.040      -0.001
chol               -0.0068      0.004     -1.840      0.066      -0.014       0.000
fbs                 0.2170      0.484      0.448      0.654      -0.732       1.166
thalachh            0.0314      0.010      3.232      0.001       0.012       0.050
exng               -1.1005      0.393     -2.802      0.005      -1.870      -0.331
caa                -0.7810      0.173     -4.525      0.000      -1.119      -0.443
===================================================================================
"""

Although the order is messed up, the logistic regression model from sklearn and from smf do have the same coefficients. The five coefficients of cp and restecg from the smf model corresponds with [1.63967367, 1.8519962, 1.72131152, 0.49135353, -0.81831285]. [-0.005895, -1.98329754, -0.02046839, -0.0067561, 0.21704038, 0.03143576, -1.10053597, -0.78101777] corresponds with the variables age through caa from the smf model. After comparing, the model from sklearn and from smf do have the same coefficients and intercepts.

    4. (18pt) With the sklearn model, let’s do some predictions on the training set (the same X and y we used to train the model from Q3):

    (6pt) The probability of having a heart attack: P(output=1|X). Print out the first 10 probabilities.

In [7]:
# Compute the probability of having a heart attack
heart_atk_prob = sklearn_logistic.predict_proba(X)
heart_atk_prob = heart_atk_prob[:, 1]
# Print the first 10 probabilities
heart_atk_prob[0:10]

array([0.73747271, 0.95009482, 0.9827361 , 0.93329596, 0.64723174,
       0.48443413, 0.92712111, 0.91447984, 0.85496029, 0.92854194])

    (6pt) The outcome labels — that is, we directly predict whether or not the person will have a heart attack, instead of predicting the probability. Print out the first 10 labels.

In [8]:
# Predict the outcome labels (whether or not a person will have heart attack)
heart_atk_label = sklearn_logistic.predict(X)
# Print the first 10 labels
heart_atk_label[0:10]

array([1, 1, 1, 1, 1, 0, 1, 1, 1, 1])

    (6pt) Show the steps of how to convert from probabilities to the labels using threshold 0.5.

In [9]:
# Convert probabilities to the labels using threshold 0.5
prob_label = np.where(heart_atk_prob > 0.5, 1, 0)
prob_label 

array([1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1,
       1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 0, 0, 1, 1, 0, 1, 0, 1, 0, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1,
       0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1,
       1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0,
       0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 1, 0, 1, 0,
       0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0,
       0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 1, 0, 0,
       1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0,

    5. (6pt) Calculate the accuracy of the predicted output labels, compared with the true output labels. You can calculate it using your own code or using sklearn.metrics package. How to interpret the accuracy? Do you think the accuracy is high enough, such that you are comfortable deploying this model in real world to predict heart attack? (Hint: you should get about 80% accuracy.)

In [10]:
# Calculate the accuracy score for the predicted output labels
accuracy = accuracy_score(y, heart_atk_label)
print('Accuracy of the predicted output labels:', accuracy)

Accuracy of the predicted output labels: 0.8217821782178217


This is interpreted as the accuracy for predicting if someone will have a heart attack or not is around 80% accurate using the model. I think the accuracy is high enough because having a heart attack is already difficult enough to predict since so many different factors can cause it. Therefore, I would be mostly comfortable deploying this model to predict heart attacks in real life. However, it is also to keep in mind that the model's predictions will not be perfect. People need to monitor and double-check before telling the patients that they will not have a heart attack.

    6. (6pt) Create a confusion matrix on the training data. Calculate accuracy, precision, and recall based on the confusion matrix.

In [11]:
# Create a confusion matrix on the training data
cm = confusion_matrix(y, heart_atk_label)
print('Confusion Matrix:')
cm

Confusion Matrix:


array([[106,  32],
       [ 22, 143]])

In [12]:
# Compute the accuracy
accuracy_cm = accuracy_score(y, heart_atk_label)
print('Accuracy:', accuracy_cm)

Accuracy: 0.8217821782178217


In [13]:
# Compute the precision
precision = precision_score(y, heart_atk_label)
print('Precision:', precision)

Precision: 0.8171428571428572


In [14]:
# COmpute the recall
recall = recall_score(y, heart_atk_label)
print('Recall:', recall)

Recall: 0.8666666666666667


# Predict AirBnB Price (40pt)

    Your second task is to predict Beijing AirBnB listing price (variable price). You use the same dataset as in PS06, and the same model (model in question 2.7).    

    1. (5pt) Replicate the model from PS06 question 2.7. Copy paste of your old code is OK.

In [15]:
# Load the airbnb data
airbnb_df = pd.read_csv('/home/jovyan/PS/data/airbnb-beijing-listings.csv', sep = ",", usecols = ["accommodates", "price", "room_type", "bedrooms", "bathrooms"])

In [16]:
# Covert price to numeric
airbnb_df['price'] = airbnb_df['price'].apply(lambda x: x.replace('$', ''))
airbnb_df['price'] = airbnb_df['price'].apply(lambda x: x.replace(',', ''))
airbnb_df['price'] = pd.to_numeric(airbnb_df['price'])

In [17]:
# Drop the entries with missing or invalid price, bedrooms or other variables
airbnb_df.dropna(inplace = True)
# 0 cannot exsists when doing log transformations, remove prices that are 0 and also price should not be 0
airbnb_df = airbnb_df[airbnb_df.price != 0]
airbnb_df.isna().sum()

room_type       0
accommodates    0
bathrooms       0
bedrooms        0
price           0
dtype: int64

In [18]:
# Create another variable called bedroom_category that is based on the number of bedrooms
airbnb_df['bedroom_category'] = pd.cut(airbnb_df.bedrooms,
                                       bins = [-np.inf, 0, 1, 2, 3, np.inf],
                                       labels = ["0", "1", "2", "3", "4+"])

In [19]:
# Create a new variable accommodates_category that contains the recoded accommodates
airbnb_df['accommodates_category'] = pd.cut(airbnb_df.accommodates,
                                            bins = [-np.inf, 1, 2, 3, np.inf],
                                            labels = ["1", "2", "3", "4 and more"])

# Create a new variable bathrooms_category that contains the recoded bathrooms
airbnb_df['bathrooms_category'] = pd.cut(airbnb_df.bathrooms,
                                         bins = [-np.inf, 0, 1, 2, np.inf],
                                         labels = ["0", "1", "2", "3 and more"])
# Create log price
log_price = np.log(airbnb_df.price)
# Create a new linear regression model with added variables
bed_bath_acc_regression = smf.ols("log_price ~ C(bedroom_category) + C(room_type) + C(accommodates_category) + C(bathrooms_category)", data = airbnb_df).fit()
bed_bath_acc_regression.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              log_price   R-squared:                       0.457
Model:                            OLS   Adj. R-squared:                  0.457
Method:                 Least Squares   F-statistic:                     2709.
Date:                Thu, 17 Mar 2022   Prob (F-statistic):               0.00
Time:                        18:52:50   Log-Likelihood:                -35751.
No. Observations:               38686   AIC:                         7.153e+04
Df Residuals:                   38673   BIC:                         7.164e+04
Df Model:                          12                                         
Covariance Type:            nonrobust                                         
==========================================================================================================
                                             coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------------------------------
Intercept                                  5.5770      0.062     90.051      0.000       5.456       5.698
C(bedroom_category)[T.1]                   0.0558      0.040      1.395      0.163      -0.023       0.134
C(bedroom_category)[T.2]                   0.1898      0.041      4.632      0.000       0.110       0.270
C(bedroom_category)[T.3]                   0.4926      0.042     11.633      0.000       0.410       0.576
C(bedroom_category)[T.4+]                  0.8730      0.044     19.909      0.000       0.787       0.959
C(room_type)[T.Private room]              -0.3243      0.007    -43.838      0.000      -0.339      -0.310
C(room_type)[T.Shared room]               -0.9453      0.017    -56.064      0.000      -0.978      -0.912
C(accommodates_category)[T.2]              0.3298      0.013     24.452      0.000       0.303       0.356
C(accommodates_category)[T.3]              0.3921      0.017     23.543      0.000       0.359       0.425
C(accommodates_category)[T.4 and more]     0.6066      0.015     39.175      0.000       0.576       0.637
C(bathrooms_category)[T.1]                 0.0049      0.047      0.106      0.916      -0.087       0.097
C(bathrooms_category)[T.2]                 0.0386      0.047      0.819      0.413      -0.054       0.131
C(bathrooms_category)[T.3 and more]        0.5938      0.049     12.017      0.000       0.497       0.691
==============================================================================
Omnibus:                     9116.538   Durbin-Watson:                   1.779
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            54083.795
Skew:                           1.003   Prob(JB):                         0.00
Kurtosis:                       8.434   Cond. No.                         52.9
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

    2. (10pt) Now use the model above to predict (log) price for each listing in your data.

In [20]:
# Use the model above to predict log price for each listing in data
y_hat = bed_bath_acc_regression.predict(airbnb_df)
print('First 5 predictions:')
print(y_hat.head())

First 5 predictions:
0    6.714846
1    5.643265
2    5.967600
3    5.967600
4    5.967600
dtype: float64


    3. (10pt) Compute root-mean-squared-error (RMSE) of this prediction. RMSE is explained in lecture notes, 4.1.5 “Model evaluation: MSE, RMSE, R2”.

In [21]:
# Compute the root-mean-sqaured-error of the prediction
rmse = mean_squared_error(log_price, y_hat) ** 0.5
print('Root-Mean-Squared-Error:', rmse)

Root-Mean-Squared-Error: 0.6096875134569459


    4. (10pt) Now use your model to predict the price for a 2-bedroom apartment that accommodates (i.e. a full 2BR apartment).

In [22]:
# Create a test to predict price with 
test = pd.DataFrame({"bedroom_category":["2"], "room_type":["Entire home/apt"], "accommodates_category":["4 and more"], "bathrooms_category":["2"]})
# Predict the log price
bed_bath_acc_regression.predict(test)

0    6.412053
dtype: float64

In the prediction for the log price, several assumptions are made. First, it is assumed that there are 2 corresponding bathrooms since there are 2 bedrooms. Second, since the question says "a 2-bedroom apartment", the room type is assumed to be entire home/apt.

    5. (5pt) Compute the average log price for all listings in this group (2BR apartment that accommodates 4). Compare the result with your prediction. How close did you get?

In [23]:
# Find other matching conditions in the original dataframe
match = airbnb_df[(airbnb_df.bedroom_category == '2') & (airbnb_df.room_type == 'Entire home/apt') & (airbnb_df.accommodates_category == '4 and more') & (airbnb_df.bathrooms_category == '2')]
# Compute the average log price for all listings in the group
np.mean(np.log(match.price))

6.509954544084154

After comparing the result with my prediction, the results are very similar. My prediction predictes that an airbnb with the conditions has a log price of 6.412053 and the averge log price for all listings with the same conditions is about 6.509955. The difference is about 0.098 which is fairly small.